qiskit_experiments/curve_analysis/grouped_curve_analysis.py

```diff
@@ -0,0 +1,324 @@
+
+    .. code-block:: python
+
+        import lmfit
+        from qiskit_experiments.curve_analysis import MultiGroupCurveAnalysis
+
+        analysis = MultiGroupCurveAnalysis(
+            groups=["groupA", "groupB"],
+            group_data_sort_key=[{"condition": "A"}, {"condition": "B"}],
+            models=[
+                lmfit.Model(fit_func_x, name="curve_X", data_sort_key={"model": "X"}),
+                lmfit.Model(fit_func_y, name="curve_Y", data_sort_key={"model": "Y"}),
+                lmfit.Model(fit_func_z, name="curve_Z", data_sort_key={"model": "Z"}),
+            ]
+        )
+
+    In above setting, each fit curve will appear in the output figure with unique name
+    in the form of ``{group name}_{model name}``, e.g. ``groupA_curve_X``.
+
+    A subclass can implement the following extra method.
+
+    .. rubric:: _create_composite_analysis_results
+
+    This method computes new quantities by taking all fit outcomes and
+    store the new values in the analysis results.
+    By default, no extra quantity is computed.
+
+    """
+
+    def __init__(
+        self,
+        groups: List[str],
+        group_data_sort_key: List[Dict[str, Any]],
+        models: Optional[List[lmfit.Model]] = None,
+    ):
+        """Create new analysis.
+
+        Args:
+            groups: List of group names.
+            group_data_sort_key: List of sort key dictionary for each group.
+            models: List of LMFIT ``Model`` class to define fitting functions and
+                parameters. Provided models are shared among groups.
+
+        Raises:
+            AnalysisError: When groups and group_data_sort_key don't match.
+        """
+        super().__init__()
+
+        if len(groups) != len(group_data_sort_key):
+            raise AnalysisError(
+                "Number of groups and data sort key doesn't match. "
+                "Data sort key must be provided for each group."
+            )
+
+        self._groups = groups
+        self._group_data_sort_key = group_data_sort_key
+        self._models = models
+
+    @classmethod
+    def _default_options(cls) -> Options:
+        """Default analysis options.
+
+        Analysis Options:
+            p0 (Union[List[Dict[str, float]], Dict[str, float]]): A dictionary of initial guesses
+                for the fit parameters keyed on the fit parameter names.
+                This can be provided for either each fit group or entire groups.
+                To provide initial guesses for each group, a list of dictionary must be
+                set here. The list ordering corresponds to the groups.
+        """
+        return super()._default_options()
+
+    @property
+    def parameters(self) -> List[str]:
+        """Return parameters of this curve analysis."""
+        unite_params = []
+        for model in self._models:
+            for name in model.param_names:
+                if name not in unite_params and name not in self.options.fixed_parameters:
```

Comment:
Would a set be better here for `unite_params`? Can always convert to a list upon return.

Reply:
Here I just respect the ordering of arguments. Using set will give you fit parameters in random ordering. If you have function `F(x, a1, a2, a3)` this will return `[a1, a2, a3]`.